src/backend/distributed/planner/multi_planner_utils.c

```diff
@@ -0,0 +1,863 @@
+
+		commonRteIdentities = bms_add_member(commonRteIdentities, rteIdentity);
+	}
+
+	/* check whether all relations exists in the main restriction list */
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+		int rteIdentity = GetRTEIdentity(relationRestriction->rte);
+
+		if (PartitionKey(relationRestriction->relationId) &&
+			!bms_is_member(rteIdentity, commonRteIdentities))
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * ReferenceRelationCount iterates over the relations and returns the reference table
+ * relation count.
+ */
+static uint32
+ReferenceRelationCount(RelationRestrictionContext *restrictionContext)
+{
+	ListCell *relationRestrictionCell = NULL;
+	uint32 referenceRelationCount = 0;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+
+		if (PartitionMethod(relationRestriction->relationId) == DISTRIBUTE_BY_NONE)
+		{
+			referenceRelationCount++;
+		}
+	}
+
+	return referenceRelationCount;
+}
+
+
+/*
+ * GenerateAttributeEquivalencesForRelationRestrictions gets a relation restriction
+ * context and returns a list of AttributeEquivalenceClass.
+ *
+ * The algorithm followed can be summarized as below:
+ *
+ * - Per relation restriction
+ *     - Per plannerInfo's eq_class
+ *         - Create an AttributeEquivalenceClass
+ *         - Add all Vars that appear in the plannerInfo's
+ *           eq_class to the AttributeEquivalenceClass
+ *               - While doing that, consider LATERAL vars as well.
+ *                 See GetVarFromAssignedParam() for the details. Note
+ *                 that we're using parentPlannerInfo while adding the
+ *                 LATERAL vars given that we rely on that plannerInfo.
+ *
+ */
+static List *
+GenerateAttributeEquivalencesForRelationRestrictions(RelationRestrictionContext
+													 *restrictionContext)
+{
+	List *attributeEquivalenceList = NIL;
+	ListCell *relationRestrictionCell = NULL;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+		List *equivalenceClasses = relationRestriction->plannerInfo->eq_classes;
+		ListCell *equivalenceClassCell = NULL;
+
+		foreach(equivalenceClassCell, equivalenceClasses)
+		{
+			EquivalenceClass *plannerEqClass =
+				(EquivalenceClass *) lfirst(equivalenceClassCell);
+
+			AttributeEquivalenceClass *attributeEquivalance =
+				AttributeEquivalenceClassForEquivalenceClass(plannerEqClass,
+															 relationRestriction);
+
+			attributeEquivalenceList =
+				AddAttributeClassToAttributeClassList(attributeEquivalenceList,
+													  attributeEquivalance);
+		}
```

Comment:
I'm still concerned with the amount of redundancy this is going to generate.  Aren't we frequently going to end up with the same thing repeated over and over here?  Yes, we're filtering some of that out later, but that's still O(n^3) in total or such.